### PR TITLE
Refactor kernels with modern loops

### PR DESCRIPTION
**Pull request by orrrrtem**
Code snippet which we will have if support modern cpp loops.

### CODE REVIEW

File: cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp [hunk 1]
```diff
@@ -20,6 +20,8 @@
 
 namespace oneapi::dal::preview::triangle_counting::backend {
 
+using namespace oneapi::dal::preview::detail;
+
 template <typename Index>
 ONEDAL_FORCEINLINE std::int64_t intersection(const Index* neigh_u,
```
Comment (andreynv) [inline cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp:26]: in this case we use neigh_v and below - neighbors - can be aligned in either way

File: cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp [hunk 2]
```diff
@@ -68,37 +70,27 @@ ONEDAL_FORCEINLINE std::int64_t intersection_local_tc(const Index* neigh_u,
...
 
     const std::int32_t average_degree_sparsity_boundary = 4;
     if (average_degree < average_degree_sparsity_boundary) {
-        dal::detail::threader_for(g_vertex_count, g_vertex_count, [&](std::int32_t u) {
-            for (auto v_ = g_vertex_neighbors + g_edge_offsets[u];
-                 v_ != g_vertex_neighbors + g_edge_offsets[u + 1];
-                 ++v_) {
-                std::int32_t v = *v_;
+        dal::detail::threader_for(vertex_count, vertex_count, [&](std::int32_t u) {
+            for (auto v: get_topology_vertex_neighbors(t, u)) {
```
Comment (andreynv) [inline cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp:88]: what impact on perf with introduction of the function call?

File: cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp [hunk 3]
```diff
@@ -68,37 +70,27 @@ ONEDAL_FORCEINLINE std::int64_t intersection_local_tc(const Index* neigh_u,
...
                     break;
                 }
-                auto u_neighbors_ptr = g_vertex_neighbors + g_edge_offsets[u];
-                for (auto w_ = g_vertex_neighbors + g_edge_offsets[v];
-                     v_ != g_vertex_neighbors + g_edge_offsets[v + 1];
-                     ++w_) {
-                    std::int32_t w = *w_;
+                auto u_neighbors_ptr = get_topology_vertex_neighbors_begin(t, u);
+                for (auto w: get_topology_vertex_neighbors(t, v)) {
                     if (w > v) {
```
Comment (andreynv) [inline cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp:94]: mostly opt topic, not relevant to style - can we iterate till v or so?

File: cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp [hunk 4]
```diff
@@ -107,68 +99,71 @@ ONEDAL_FORCEINLINE array<std::int64_t> triangle_counting_local_(
...
     }
     else { //average_degree >= average_degree_sparsity_boundary
-        dal::detail::threader_for_simple(g_vertex_count, g_vertex_count, [&](std::int32_t u) {
-            if (g_degrees[u] >= 2)
+        dal::detail::threader_for_simple(vertex_count, vertex_count, [&](std::int32_t u) {
+            if (get_topology_vertex_degree(t, u) >= 2)
                 dal::detail::threader_for_int32ptr(
-                    g_vertex_neighbors + g_edge_offsets[u],
-                    g_vertex_neighbors + g_edge_offsets[u + 1],
+                    get_topology_vertex_neighbors_begin(t, u),
```
Comment (andreynv) [inline cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp:115]: the only minor concern - too long name of the function

File: cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp [hunk 5]
```diff
@@ -107,68 +99,71 @@ ONEDAL_FORCEINLINE array<std::int64_t> triangle_counting_local_(
...
                         if (v <= u) {
-                            const std::int32_t* neigh_u = g_vertex_neighbors + g_edge_offsets[u];
-                            std::int32_t size_neigh_u =
-                                g_vertex_neighbors + g_edge_offsets[u + 1] - neigh_u;
-                            const std::int32_t* neigh_v = g_vertex_neighbors + g_edge_offsets[v];
-                            ;
-                            std::int32_t size_neigh_v =
-                                g_vertex_neighbors + g_edge_offsets[v + 1] - neigh_v;
-                            std::int32_t new_size_neigh_v;
+                            const std::int32_t* v_neighbors_ptr =
```
Comment (andreynv) [inline cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp:120]: we switched to int32 here, why

File: cpp/oneapi/dal/graph/detail/undirected_adjacency_vector_graph_impl.hpp [hunk 6]
```diff
@@ -227,14 +227,26 @@ constexpr auto get_topology_vertex_degree(const topology<IndexType>& _topology,
     return _topology._degrees[vertex];
 }
 
+template <typename IndexType>
```
Comment (andreynv) [inline cpp/oneapi/dal/graph/detail/undirected_adjacency_vector_graph_impl.hpp:230]: here we use 'IndexType' while earlier I saw Index - align

File: cpp/oneapi/dal/graph/detail/undirected_adjacency_vector_graph_impl.hpp [hunk 7]
```diff
@@ -227,14 +227,26 @@ constexpr auto get_topology_vertex_degree(const topology<IndexType>& _topology,
...
+constexpr auto get_topology_vertex_neighbors_end(const topology<IndexType>& _topology,
+                                                 const IndexType& vertex) noexcept ->
+    typename topology<IndexType>::const_vertex_edge_iterator {
+    return _topology._cols.get_data() + _topology._rows[vertex + 1];
+}
+
 template <typename IndexType>
 constexpr auto get_topology_vertex_neighbors(const topology<IndexType>& _topology,
                                              const IndexType& vertex) noexcept ->
     typename topology<IndexType>::const_vertex_edge_range {
```
Comment (andreynv) [inline cpp/oneapi/dal/graph/detail/undirected_adjacency_vector_graph_impl.hpp:247]: the use of constness - to discuss separately

File: cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp (hunk 1 above)
Comment (orrrrtem) [reply cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp:26]: This was not aligned. The only aligned thing as part of demonstration is triangle_counting_local_

File: cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp (hunk 3 above)
Comment (orrrrtem) [reply cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp:94]: I see no ways here. The place of v is unpredictable as graph is in CSR format.

File: cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp (hunk 4 above)
Comment (orrrrtem) [reply cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp:115]: Yes. Name can be shorter. We can remove topology and to have the same name of service as the public graph service beside used namespace detail.

File: cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp (hunk 5 above)
Comment (orrrrtem) [reply cpp/oneapi/dal/algo/triangle_counting/backend/cpu/vertex_ranking_default_kernel_scalar.hpp:120]: It is layout field type. auto can be used here. Also, we cane reimplement the loop used this ptr with iterators:
                            std::int32_t new_v_neighbors_count = 0;
                            for (auto w: get_topology_vertex_neighbors(t, v)) {
                                if (w >= v) {
                                    break;
                                }
                                new_v_neighbors_count++;
                            }

File: cpp/oneapi/dal/graph/detail/undirected_adjacency_vector_graph_impl.hpp (hunk 6 above)
Comment (orrrrtem) [reply cpp/oneapi/dal/graph/detail/undirected_adjacency_vector_graph_impl.hpp:230]: To align later after coding extension will be designed finally.

**Pull request closed without merging**